### PR TITLE
Accept unpacked pre-routed topk in fp8 block-scale and bf16 MoE

### PR DESCRIPTION
**Pull request by b8zhong**
`trtllm_fp8_block_scale_routed_moe` and `trtllm_bf16_routed_moe` only take the packed `(expert_id << 16) | bf16(weight)` form, which truncates the fp32 routing weights callers already have. However FP4 already handles this.

#3763 gave the FP4 path`(topk_ids, topk_weights)` form; this does the same for the FP8 block-scale and BF16 runners.

And unified support in https://github.com/flashinfer-ai/flashinfer/pull/4104

Now fp32 weights reach the combine as fp32.

## Summary by CodeRabbit

- **New Features**
  - Added logits-based, packed precomputed, and unpacked precomputed routing for BF16, FP8, and FP4 Mixture-of-Experts operations.
  - Routed APIs now accept packed expert IDs or separate expert IDs and weights.
  - Precomputed routing supports FP32 and BF16 expert weights where applicable.

- **Bug Fixes**
  - Improved validation and handling of unpacked routing inputs.

- **Tests**
  - Added FP8 routing coverage for multiple expert-weight formats.

### CODE REVIEW

File: csrc/trtllm_fused_moe_kernel_launcher.cu [hunk 1]
```diff
@@ -784,16 +808,23 @@ class Bf16MoeLauncher : public FusedMoeLauncher {
...
     }
+    if (is_unpacked_routing()) {
+      TVM_FFI_ICHECK(expert_weights.dtype() == dl_bfloat16 || expert_weights.dtype() == dl_float32)
+          << "expert_weights must be bfloat16 or float32 for unpacked precomputed routing.";
+      TVM_FFI_ICHECK_EQ(expert_weights.ndim(), 2) << "expert_weights must be 2D.";
+      TVM_FFI_ICHECK_EQ(expert_weights.size(0), hidden_states.size(0))
+          << "expert_weights and hidden_states must have same number of tokens.";
+      TVM_FFI_ICHECK_EQ(expert_weights.size(1), args->top_k)
+          << "expert_weights dim1 must match top_k.";
+    }
```
[the remark this answers is not in the document]
Comment (b8zhong) [reply csrc/trtllm_fused_moe_kernel_launcher.cu:830]: Fixed.

File: flashinfer/fused_moe/core.py [hunk 2]
```diff
@@ -2005,6 +2008,7 @@ def trtllm_bf16_moe_op(
             tuning_config,
             moe_inputs.to_list(),
             routing_bias=routing_bias,
+            routing_input_mode=routing_input_mode,
```
[the remark this answers is not in the document]
Comment (b8zhong) [reply flashinfer/fused_moe/core.py:2012]: Fixed.

File: flashinfer/fused_moe/core.py [hunk 3]
```diff
@@ -3524,9 +3545,11 @@ def trtllm_bf16_routed_moe(
 
     Parameters
     ----------
-    topk_ids : torch.Tensor
+    topk_ids : torch.Tensor or Tuple[torch.Tensor, torch.Tensor]
         ``[seq_len, top_k]`` int32 tensor of packed expert indices and
         weights.  Format ``(expert_id << 16) | (weight_bf16.view(int16))``.
+        Alternatively a ``(topk_ids, topk_weights)`` pair of plain ``int32``
+        indices and ``bfloat16`` or ``float32`` weights.
```
[the remark this answers is not in the document]
Comment (b8zhong) [reply flashinfer/fused_moe/core.py:3555]: Fixed.

File: tests/moe/test_trtllm_gen_fused_moe.py [hunk 4]
```diff
@@ -1886,6 +1886,37 @@ def test_fp8_block_scale_routed_activation_type_relu2_smoke():
...
+            weight_layout=WeightLayout.MajorK.value,
+            enable_pdl=True,
+            fp8_quantization_type=fp8_quantization_type,
+            activation_type=activation_type,
+        ).to(torch.float)
+        close = torch.isclose(output_ref, output_unpacked, atol=1e-2, rtol=1e-2)
+        mismatch_pct = (~close).float().mean().item() * 100
+        assert mismatch_pct < 10, (
+            f"{weights.dtype} unpacked mismatch percentage is {mismatch_pct:.2f}%"
+        )
```
[the remark this answers is not in the document]
Comment (b8zhong) [reply tests/moe/test_trtllm_gen_fused_moe.py]: Fixed.

**Comment by b8zhong** [comment]
Hi @Aneureka , I'm wondering if you could help review this PR, as you reviewed the previous one fixing the NVFP4 case. This one fixes the FP8 and BF16 one. Thanks!

**Comment by Aneureka** [comment]
LGTM overall and it does not break user-faced API. Let's wait for the test pipeline.
@aleozlx may take a look as well.

**Review by Aneureka — APPROVED** [review]

**Comment by b8zhong** [comment]
Thanks! Please let me know if there is internal CI failures.

**Comment by b8zhong** [comment]
@Aneureka It appears both internal and public is green. I'm wondering if we should merge this.

**Comment by Aneureka** [comment]
Sure. Let's merge it.
